### ISSUE TITLE
Placing a Mach port into multiple port sets doesn't work

### ISSUE THREAD

**Issue by LubosD** [darlinghq/darling#368]
It can be placed there, but then only one of these portsets will get woken up upon port activity.

**Comment by LubosD** [comment]
Test app that triggers the bug:
[multipset-example.tar.gz](https://github.com/darlinghq/darling/files/2313273/multipset-example.tar.gz)

**Comment by bugaevc** [comment]
A backstory for people who follow along: we traced this all the way through the stack from buttons not handling mouse clicks properly.

A control like `NSButton` has basically two ways of handling events sequences like "mouse down, mouse move, mouse up", either by remembering the fact of `mouseDown` and returning back to the event loop, like this:

```objc
- (void) mouseDown: (NSEvent *) event {
    _mousePressed = YES;
    [self setNeedsDisplay: YES];
}

- (void) mouseUp: (NSEvent *) event {
    if (_mousePressed) {
        // handle the click
        [self sendAction: [self action] to: [self target]];
    }
    _mousePressed = NO;
    [self setNeedsDisplay: YES];
}
```

Alternatively, it can run a nested run loop inside `mouseDown:` like this:

```objc
- (void) mouseDown: (NSEvent *) event {
    _mousePressed = YES;
    [self setNeedsDisplay: YES];
    event = [[self window] nextEventMatchingMask: NSLeftMouseUpMask];
    // handle the click
    [self sendAction: [self action] to: [self target]];
    _mousePressed = NO;
    [self setNeedsDisplay: YES];
}
```

See [this doc](https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/EventOverview/HandlingMouseEvents/HandlingMouseEvents.html#//apple_ref/doc/uid/10000060i-CH6-SW18) for more details.

By default, `NSControl` [uses](https://github.com/darlinghq/darling-cocotron/blob/master/AppKit/NSControl.m#L571-L610) the second approach, running a nested run loop. It uses both `-[NSCell trackMouse:inRect:ofView:untilMouseUp:untilMouseUp]` (which [wraps](https://github.com/darlinghq/darling-cocotron/blob/master/AppKit/NSCell.m#L809-L859) `-[NSWindow nextEventMatchingMask:]`) and `-[NSWindow nextEventMatchingMask:]` itself.

Next, `-[NSWindow nextEventMatchingMask:]` ends up running `[NSRunLoop currentRunLoop]` in `NSEventTrackingRunLoopMode`. `NSRunLoop` is a [thin wrapper](https://github.com/darlinghq/darling-foundation/blob/master/src/NSRunLoop.m) over [`CFRunLoop`](https://github.com/darlinghq/darling-corefoundation/blob/master/CFRunLoop.c). Each run loop mode corresponds to a Mach port set that loops listens on when run in this mode.

The bug we've been seeing is clicking a button would hang the process forever; `nextEventMatchingMask:` never returned. What happened was the run loop was never woken up by the X11 socket becoming readable. The way _that_ is supposed to work is there is a separate thread, `__CFSocketManagerThread`, that [gets spawned](https://github.com/darlinghq/darling-corefoundation/blob/master/CFSocket.c#L2474-L2483) the first time you add a `CFSocket` to a run loop; this thread [`select()`s on the Unix fds](https://github.com/darlinghq/darling-corefoundation/blob/master/CFSocket.c#L2151) the loop should listen on, and as soon as one becomes ready, the socket manager threads sends a Mach message to the run loop _wakeup port_ (then the loop wakes up and services that `CFSocket`).

So whenever a new run loop mode is initialized, the wakeup port of that run loop (a loop only has one wakeup port) would [get inserted](https://github.com/darlinghq/darling-corefoundation/blob/master/CFRunLoop.c#L816) into new mode's port set. And if there are multiple modes, the same wakeup port would get inserted into each port set.

There aren't a lot of docs that mention inserting a port into multiple sets, and ones that exist contradict each other (& at times themselves). Some say that:
> a port can only belong to one portset at once

others:
> If the receive right is already a member of another port set, that relationship is unafected by this operation.  A receive right can be in multiple port sets simultaneously.

Currently on Darling, _inserting_ a port into multiple port sets succeeds (with `KERN_SUCCESS`), but it doesn't _work_, i.e. a thread listening for messages sent to a port set (one that the port was inserted into, but not the first one of those) doesn't get woken up when a message is sent to the _port_. Presumably, this is because the XNU code _does_ support having a port in multiple port sets, but our Linux duct tape code doesn't account for the case that there be multiple port sets this port is a member of so there may be multiple threads that it needs to wake up.

So for the `CFRunLoop` & X11 socket case it means that waking up the run loop on the socket becoming readable only works if the run loop runs in the `NSDefaultRunLoopMode` (aka `kCFRunLoopDefaultMode`), which is why the events were never delivered to the app, which is why the button wouldn't respond.

As a workaround until the kernel bug is fixed, I build AppKit with `NSEventTrackingRunLoopMode` and other run loop mode names changed to equal `NSDefaultRunLoopMode` (it's not enough for them to be `CFSTR("kCFRunLoopDefaultMode")`, they really need to point to that same `CFString` object; this also causes headache in Swift, where strings are implicitly bridged to Swift's native `String` type, so it's harder to keep object identity where needed, but that's another story).

To summarize, that's an interesting and rare case where a bug in the LKM manifested as a problem with the UI, having to do with Mach ports, Unix sockets, X11, threads and event loops. Wow.

**Comment by LubosD** [comment]
The reason is how "hacked" `ipc_mqueue_post` is in Darling.

I think it may be easier to finish the xnu-upgrade branch work where the intention is NOT to have the whole waiting system modified. (The XNU waiting system is coincidentally also completely overhauled in that branch.)

**Comment by LubosD** [comment]
The test program now works in the branch for issue #275.

**Comment by LubosD** [comment]
I consider the bug resolved - in the vchroot branch. Will be merged into master as soon as we make sure there are no major regressions.
